### ISSUE TITLE
`saved_tensors_hooks`/`_pop_saved_tensors_default_hooks` API extension

### ISSUE THREAD

**Issue by albertz** [pytorch/pytorch#129867]
### 🚀 The feature, motivation and pitch

I want to implement gradient checkpointing API where I want to specify some tensors which are not stored for backpropagation. The current `checkpoint` API does not allow for that. Only intermediate tensors are not stored, but the final tensor will likely be stored, as this would be outside of what `checkpoint` handles.

Related discussions:
* https://discuss.pytorch.org/t/gradient-checkpointing/205416
* https://github.com/rwth-i6/returnn/issues/1552

To implement what I want, I however need to extend the `saved_tensors_hooks` scope up to a future point once the relevant tensor goes out of scope. I can of course do that for some specific models, but I don't really see any good way to do this in a general way. For that, the `saved_tensors_hooks` API somehow would need to be extended.

My current very hacky solution is:

* Manually call `saved_tensors_hooks.__enter__` when I need to enter.
* Detect when the relevant `tensor`s go out of scope. At that point, call `saved_tensors_hooks.__exit__`.

This hacky solution is very problematic for multiple reasons:

* Detecting when the tensors go out of scope is tricky.
* Calling `saved_tensors_hooks.__exit__` from a different thread does not work because the scope is thread-local. (E.g. when I put some logic in `__del__` somewhere to do this, I can potentially run into this problem.)
* Calling `saved_tensors_hooks.__exit__` when there was another unrelated `saved_tensors_hooks.__enter__` in the meantime will lead to wrong behavior.

Also, actually a more fine-grained hook would be even nicer: Basically whenever `pack_hook` is called on my relevant tensors, I want to execute my hook, but otherwise not.

cc @soulitzer

**Comment by soulitzer** [comment]
Thanks for filing this issue. The proposal in this issue is interesting, but one issue is that depending on when those tensors end up actually being saved, it can arbitrarily extend the hooks region. 

Along the lines of this proposal, one can imagine a different approach where you (1) introduce a new API that allows one to mark a tensor to run hooks upon saving, and then (2) extend torch.utils.checkpoint to mark its outputs using this API, (3) save some additional AC state to remember what additional buffers need to be packed during recompute. This would ensure that the outputs of AC are not saved, while avoiding potentially recomputing extra operations.

With either approach though, I feel like we're adding a lot of complexity while avoiding solving a more fundamental problem.

Checkpointing today operates on the granularity of tensors being saved for backward because it interposes on the pack/unpack operations. 

This means that it is very natural to build a checkpoint today that says, "for all tensors SAVED in this region, I will recompute them". But (what this issue is pointing out is that) what one might want say instead is "for all tensors CREATED in this region, I will recompute them". This may be preferred in the sense that as a user I know what operations are cheap, and what outputs are expensive, and therefore I am okay with recomputing this specific operation. When operations are saved, on the other hand, is bit more hidden and not as intuitive.

Selective checkpointing somewhat gets at this issue, but again, due to the way checkpoint is implemented, there is some friction there that seem hard to solve fundamentally, e.g. a policy function that defaults to saving everything saves outputs of all operations, not only tensors that are needed for backward (I would note that this is less of a problem in compile however).

I think a better way of approaching this type of issue is to explore a new checkpoint-like API instead that is based primarily on torch dispatch. Copy-pasting [link](https://discuss.pytorch.org/t/gradient-checkpointing/205416/8?u=soulitzer) here for visibility.

I'm not sure how much time I would have in the very near term to look at this, but if you still blocked on this issue, I would recommend exploring this approach some more as it should be able to be done entirely in user-code.

**Comment by albertz** [comment]
> one issue is that depending on when those tensors end up actually being saved, it can arbitrarily extend the hooks region.

Yes. But I actually also only want to have the hook for some selected tensors, not for all. So that leads exactly to such different API as you describe.

> This means that it is very natural to build a checkpoint today that says, "for all tensors SAVED in this region, I will recompute them". But (what this issue is pointing out is that) what one might want say instead is "for all tensors CREATED in this region, I will recompute them".

I'm a bit confused by your terminology here. Maybe you mean the same as what I want to achieve but I wanted to clarify this part here. What does it mean for a tensor to be "saved" vs a tensor to be "created"?

Maybe take my example (from https://github.com/rwth-i6/returnn/issues/1552#issuecomment-2195504210) for the following computation:
```python
x = a + b
y = x * c
```
I want an API where I can specify that `x` is not saved but recomputed. I suggested a gradient checkpoint API which looks like this:
```python
with gradient_checkpoint_scope():
    x = a + b
y = x * c
```
Implementing `gradient_checkpoint_scope` via `saved_tensors_hooks` is not easily possible because `saved_tensors_hooks` would need to go beyond the scope of `gradient_checkpoint_scope` because `x` is used afterwards, and I need to hook pack/unpack for all/any future uses of `x`.

With the current API of `checkpoint`, you would need to put the scope around the whole code for this example, like:
```python
y = checkpoint(lambda: (a + b) * c)
```
So the multiplication with `c` must be part of the function. Basically any other use of `x` must be part of the function. The problem is, I don't know exactly where `x` is maybe used.

So yes, `x` is created in this `gradient_checkpoint_scope` scope, and I want to specify that `x` should not be saved (no matter what comes outside the scope) but recomputed.

> This may be preferred in the sense that as a user I know what operations are cheap, and what outputs are expensive, and therefore I am okay with recomputing this specific operation.

Yes exactly. `x = a + b` is a cheap op, and I know that `a` will anyway be stored (it's a `torch.nn.Parameter` in my real-world example), so I just want to avoid storing `x`, as this would cost too much memory.

> When operations are saved, on the other hand, is bit more hidden and not as intuitive.

I'm not exactly sure I understand the difference here. When I write `x = a + b`, and you say to save the operation, what is saved here?



> I think a better way of approaching this type of issue is to explore a new checkpoint-like API instead that is based primarily on torch dispatch. Copy-pasting [link](https://discuss.pytorch.org/t/gradient-checkpointing/205416/8?u=soulitzer) here for visibility.

That's also what I linked in the issue. But I think this is orthogonal. Torch dispatch allows that there is no explicit need for the user to define an explicit function, which is then called again, but instead, it records the computation graph, and replays those parts which are necessary, i.e. which should be recomputed. I'm also very much in favor of that, as this feels much simpler. This allows to define a scope instead of defining such a function.

Unless you use `saved_tensors_hooks` around the whole model code. The torch dispatch can still be around your defined scope (`x = a + b`), when you know that only those operations/tensors should be recomputed, i.e. you only need to record the computation graph of those. But when you have `saved_tensors_hooks` around everything, then you can very easily check just always whether the tensor should be stored or you can point to such recomputation logic.

But I think you want to avoid having `saved_tensors_hooks` always enabled, as this would introduce a too big performance penalty?

So that is exactly what this issue here is about, that I don't want to enable `saved_tensors_hooks` all the time, but also not just for a given code scope (because I cannot really know exactly the code scope), but instead for some selected tensors, i.e. whenever `save_for_backward` is called for some of my selected tensors.

Btw, maybe to extend a bit my simple example, to see my actual use case: I want to use this e.g. for weight noise or weight dropout. The weight (`torch.nn.Parameter`) is anyway stored. The dropout mask can be recomputed when I store the RNG state, and then I can recompute the result on-the-fly. I want to implement this in a generic way that I can use such weight dropout anywhere, without the need to put an outer scope around it or modify the other code (e.g. modify any code where this weight is used).

Note, I'm also currently working on such an implementation for `gradient_checkpoint_scope`: https://github.com/rwth-i6/returnn/pull/1559

I'm using torch dispatch to record the computation graph, just as @soulitzer suggested. This is only for the scope under `gradient_checkpoint_scope`, because that are the ops/tensors I want to recompute. However, then I'm using `saved_tensors_hooks` in a quite hacky way: This goes beyond the scope of `gradient_checkpoint_scope`, because I want to apply the pack/unpack hook for any future use of `x` (in the example) (or in general: any tensors created under the `gradient_checkpoint_scope` scope). But then I want to close the `saved_tensors_hooks` scope at some later point, when I detect that it is not needed anymore (when `x` gets GCed). But this is hacky and complicated, and I think there could be a better API in PyTorch which makes it easier. E.g. which allows me to have the hook on specified tensors, not for some code scope.

**Comment by albertz** [comment]
Note, our implementation in RETURNN is ready now. PR https://github.com/rwth-i6/returnn/pull/1559, [code here]( https://github.com/rwth-i6/returnn/blob/master/returnn/torch/util/gradient_checkpoint.py) ([test case](https://github.com/rwth-i6/returnn/blob/a0f8be5de6b7790ae99725e5935d1b81941c2b7d/tests/test_torch_util.py#L18)).
